.github/workflows/test.yml

```diff
@@ -30,34 +30,71 @@ jobs:
 
       - run: go version
 
-      # Checks-out repository under $GITHUB_WORKSPACE
-      - name: Checkout
-        if: ${{!startsWith(matrix.os, 'ubuntu')}}
-        uses: actions/checkout@v4
-
       # Checks-out repository under $GITHUB_WORKSPACE with tags and history (needed by "SonarCloud Scan" step)
       - name: Full checkout
-        if: startsWith(matrix.os,'ubuntu')
         uses: actions/checkout@v4
         with:
           fetch-depth: 0 # Full clone for SonarCloud
 
       # Build everything
       - name: Run a build
         run: go build ./...
+        shell: bash
 
       # Runs a single command using the runners shell, -p1 for `race: limit on 8128 simultaneously alive goroutines is exceeded, dying` at macos
       - name: Run a test
-        run: go test -v -race ./... -coverpkg=./... -covermode=atomic -coverprofile=./coverage.txt
+        run: go test -v -race ./... -coverpkg=./... -covermode=atomic -coverprofile=./coverage.txt -json > ./report.json
+        shell: bash
+      
+      - name: Dump test report
+        if: always()
+        run: cat ./report.json
+        shell: bash
+
+      - name: Prepare upload files
+        run: |
+          mkdir -p ./outputs
+          cp ./coverage.txt ./outputs/${{ matrix.os }}.coverage.txt
+          cp ./report.json ./outputs/${{ matrix.os }}.report.json
+        shell: bash
+
+      - name: Upload coverage and report files
+        uses: actions/upload-artifact@v3
+        with:
+          name: ${{ hashFiles('./outputs') || 'none' }}
+          path: ./outputs
+          retention-days: 1
+          if-no-files-found: warn
+  
+  coverage-sonar-cloud-scan:
+    needs: test
+    # The type of runner that the job will run on
+    runs-on: ubuntu-latest
+    steps:
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0 # Shallow clones should be disabled for a better relevancy of analysis
+
+      - name: Download artifacts
+        uses: actions/download-artifact@v3
+        with:
+          path: ./outputs
+
+      - name: Prepare coverage and report files
+        run: |
+          mkdir -p .tmp/coverage
+          mkdir -p .tmp/report
+          find ./outputs -name "*.coverage.txt" -exec sh -c 'cp $1 .tmp/coverage/$(echo $1 | sed "s/[\/.]/-/g" ).coverage.txt' _ {}  \;
```

Comment:
From the job logs, the file name are kinda weird ("--outputs" prefix and "-coverage-txt.coverage.txt" suffix)

`/github/workspace/.tmp/coverage/--outputs-608abf8a79c6238b98b3e145b34057c18dd699466c788dc7ff714116d607577d-ubuntu-latest-coverage-txt.coverage.txt`

but if the coverage calculation works then who cares. :D